docs/async-await-proposal.md

```diff
@@ -0,0 +1,442 @@
+in terms of NIO types. Instead they now implement an `async` function.
+
+> 2. There is a different context type passed to the user function for each
+>   different type of call and this context is generic over the response type.
+
+The same non-generic `AsyncServerCallContext` is passed to the user function
+regardless of the type of RPC.
+
+> 3. In the server- and bidirectional-streaming call handlers, an added layer of
+>   asynchrony is exposed. That is, the user must return a _future_ for
+>   a closure that will handle incoming events.
+
+The user function consumes requests from an `AsyncSequence`, using the new
+language idioms.
+
+> 4. The user _must_ fulfil the `statusPromise` when it receives `.end` but there
+>   is nothing that enforces this.
+
+The user need simply return from the function or throw an error. The closing of
+the call is handled by the library.
+
+If the user function throws a `GRPCStatus` (which already conforms to `Error`)
+or a value of a type that conforms to `GRPCStatusTransformable` then the library
+will take care of setting the RPC status appropriately. If the user throws
+anything else then the library will still take care of setting the status
+appropriately, but in this case it will use `internalError` for the RPC status.
+
+### Proposed client API
+
+We propose generating a client which conforms to this new protocol:
+
+```swift
+public protocol Echo_AsyncEchoClientProtocol: GRPCClient {
+	var serviceName: String { get }
+	var interceptors: Echo_EchoClientInterceptorFactoryProtocol? { get }
+
+	func makeGetCall(
+		_ request: Echo_EchoRequest,
+		callOptions: CallOptions?
+	) -> AsyncUnaryCall<Echo_EchoRequest, Echo_EchoResponse>
+
+	func makeExpandCall(
+		_ request: Echo_EchoRequest,
+		callOptions: CallOptions?
+	) -> AsyncServerStreamingCall<Echo_EchoRequest, Echo_EchoResponse>
+
+	func makeCollectCall(
+		callOptions: CallOptions?
+	) -> AsyncClientStreamingCall<Echo_EchoRequest, Echo_EchoResponse>
+
+	func makeUpdateCall(
+		callOptions: CallOptions?
+	) -> AsyncBidirectionalStreamingCall<Echo_EchoRequest, Echo_EchoResponse>
+}
+```
+
+Here is an example use of the new client API, making a bidirectional streaming
+call to `update`:
+
+```swift
+// No longer provide a response handler when making the call.
+let update = client.makeUpdateCall()
+
+Task {
+  // Send requests as before but using `await` instead of a `promise`.
+  for word in ["foo", "bar", "baz"] {
+    try await update.sendMessage(.with { $0.text = word })
+  }
+  // Close the request stream, again using `await` instead of a `promise`.
+  try await update.sendEnd()
+}
+
+// Consume responses as an AsyncSequence.
+for try await response in update.responseStream {
+  print("update received: \(response.text)")
+}
+
+// Wait for the call to terminate, but using `await` rather than a future.
+let status = await update.status
+print("update completed with status: \(status.code)")
+```
```

Comment:
This all looks good 👍 

Not in love with `sendEnd()` I guess (one can miss sending it), but it matches the general style of the stream writer so I guess it's fine 👍

Reply:
Yep, this is why we provide the "simpler" API variants below as you already noticed in https://github.com/grpc/grpc-swift/pull/1231#discussion_r678802185.